vault/external_tests/seal_migration/seal_migration_test.go

```diff
@@ -0,0 +1,517 @@
+
+	cluster.EnsureCoresSealed(t)
+
+	return transitSeal
+}
+
+func unsealMigrate(t *testing.T, client *api.Client, keys [][]byte, transitServerAvailable bool) {
+
+	for i, key := range keys {
+
+		// Try to unseal with missing "migrate" parameter
+		_, err := client.Sys().UnsealWithOptions(&api.UnsealOpts{
+			Key: base64.StdEncoding.EncodeToString(key),
+		})
+		if err == nil {
+			t.Fatal("expected error due to lack of migrate parameter")
+		}
+
+		// Unseal with "migrate" parameter
+		resp, err := client.Sys().UnsealWithOptions(&api.UnsealOpts{
+			Key:     base64.StdEncoding.EncodeToString(key),
+			Migrate: true,
+		})
+
+		if i < keyThreshold-1 {
+			// Not enough keys have been provided yet.
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			if transitServerAvailable {
+				// The transit server is running.
+				if err != nil {
+					t.Fatal(err)
+				}
+				if resp == nil || resp.Sealed {
+					t.Fatalf("expected unsealed state; got %#v", resp)
+				}
+			} else {
+				// The transit server is stopped.
+				if err == nil {
+					t.Fatal("expected error due to transit server being stopped.")
+				}
+			}
+			break
+		}
+	}
+}
+
+// verifyBarrierConfig verifies that a barrier configuration is correct.
+func verifyBarrierConfig(t *testing.T, cfg *vault.SealConfig, sealType string, shares, threshold, stored int) {
+	t.Helper()
+	if cfg.Type != sealType {
+		t.Fatalf("bad seal config: %#v, expected type=%q", cfg, sealType)
+	}
+	if cfg.SecretShares != shares {
+		t.Fatalf("bad seal config: %#v, expected SecretShares=%d", cfg, shares)
+	}
+	if cfg.SecretThreshold != threshold {
+		t.Fatalf("bad seal config: %#v, expected SecretThreshold=%d", cfg, threshold)
+	}
+	if cfg.StoredShares != stored {
+		t.Fatalf("bad seal config: %#v, expected StoredShares=%d", cfg, stored)
+	}
+}
+
+// initializeShamir initializes a brand new backend storage with Shamir.
+func initializeShamir(
+	t *testing.T, logger hclog.Logger,
+	storage teststorage.ReusableStorage, basePort int) (string, [][]byte) {
+
+	var baseClusterPort = basePort + 10
+
+	// Start the cluster
+	var conf = vault.CoreConfig{
+		Logger: logger.Named("initializeShamir"),
+	}
+	var opts = vault.TestClusterOptions{
+		HandlerFunc:           vaulthttp.Handler,
+		NumCores:              numTestCores,
+		BaseListenAddress:     fmt.Sprintf("127.0.0.1:%d", basePort),
+		BaseClusterListenPort: baseClusterPort,
+	}
+	storage.Setup(&conf, &opts)
+	cluster := vault.NewTestCluster(t, &conf, &opts)
+	cluster.Start()
+	defer func() {
+		storage.Cleanup(t, cluster)
+		cluster.Cleanup()
+	}()
+
+	leader := cluster.Cores[0]
+	client := leader.Client
+
+	// Unseal
+	if storage.IsRaft {
+		testhelpers.RaftClusterJoinNodes(t, cluster)
+		if err := testhelpers.VerifyRaftConfiguration(leader, numTestCores); err != nil {
+			t.Fatal(err)
+		}
+	} else {
+		cluster.UnsealCores(t)
+	}
+	testhelpers.WaitForNCoresUnsealed(t, cluster, numTestCores)
```

Comment:
It seems like this would time out if the cluster's storage is Raft since we're not unsealing for that case.

Reply:
Wrote too quick, looks like `raftClusterJoinNodes` also performs unsealing.